### ISSUE TITLE
Different styling for different lines, how to separate odd and even lines?

### ISSUE THREAD

**Issue by kiwitotoro** [atom/atom#4829]
I have [asked](http://stackoverflow.com/questions/27729734/styling-alternating-lines-in-atom-editor) this on Stack overflow but it seems to be harder than I thought.

Problem: I wanted to make the odd lines a different colour so that when I set soft line wraps I can see clearly that it is one line split over 2 or 3 without having to focus on the dots in the gutter.

Attempt to fix: In my style.less I tried adding:

.editor .line:nth-child(2n){
  background-color:#292929;
}

Then I tried:
.editor .line[data-screen-row]:nth-child(2n) {
  background-color:#292929;
}

Here is the screen shot in the SO post:
![atom_line_mess](https://cloud.githubusercontent.com/assets/775993/5596890/e0cb89a2-92db-11e4-9aae-20ab755224c8.png)

The issue: The order of the lines is not related to the actual order in the document so this was not consistent. I don't know why the data-screen-row didn't work but it still did not highlight correctly.

I am currently looking into plugins but I can not figure out how to effect these line generators (if someone can tell me where to look I can work on a plugin.

I think it would need something like for example an odd and even class on the lines (It would also need to know if the line is wrapped and make it even or odd based on the original document and not the display in the DOM.

Is there a way to do this with just CSS/LESS that I missed?

Another issue I noticed is that when I add the background colour to the lines highlighting does not work, It would also be nice to add a way to control the depth so advice about how that would work would also be cool.

Thanks in advance for any advice you can offer.

**Comment by lee-dohm** [comment]
@simurai Got any ideas on this one?

**Comment by simurai** [comment]
@kiwitotoro I guess you would like "zebra stripes", but not when soft-wrapped. There it should continue with the same background color, so that it's easier to see the soft wrapped part, right? Like this:

![screen shot 2015-01-03 at 5 13 38 pm](https://cloud.githubusercontent.com/assets/378023/5601934/414c4632-936c-11e4-80bf-32049d910bf2.png)

I don't think it's possible with CSS/LESS alone. Because there is nothing that you could tell that a line is soft-wrapped. Plus the issue with the order.

Yeah, a package that adds classes might work. Since it adds the dots instead of the line numbers, somewhere there must already be that logic. But not sure where/how you can hook into that.

**Comment by kiwitotoro** [comment]
@simurai Thanks, yes zebra stripes is what I was looking for.

I am starting to build a plugin, could be a while as I do not know coffee script and the documentation is not easy to work with.

This page for example: https://atom.io/docs/api/v0.165.0/TextEditor says I should use `TextEditorView ` for dealing with the visual appearance of editors but it goes to a 404...

I have got the following:

    atom.workspace.observeTextEditors (editor) ->
      console.log(editor.getLineCount());

To show the number of lines in each open editor. (I don't know how to get only the active tab)

Also this (while very fragile looking) seems to give me the rendered lines:

    buf = editor.getLineDecorations();
      console.log(buf[0].displayBuffer.screenLines);

So where Line 9 is the first of a soft wrapped line and 10 is its overflow:

    9: Tokenized
      LinebufferDelta: 28
      fold: undefined
      id: 151
      indentLevel: undefined
      invisibles: null
      lineEnding: null
      ruleStack: Array[1]
      startBufferColumn: 0
      tabLength: undefined
      text: "    @zebraStripesView = new "
      tokens: Array[8]
      __proto__: TokenizedLine
    10: TokenizedLine
      bufferDelta: 45
      fold: undefined
      id: 152
      indentLevel: undefined
      invisibles: null
      lineEnding: "↵"
      ruleStack: Array[1]
      startBufferColumn: 28
      tabLength: undefined
      text: "ZebraStripesView(state.zebraStripesViewState)"
      tokens: Array[6]
      __proto__: TokenizedLine

I could loop through lines and tell if they are not the last if they have an empty `lineEnding` value. I would still need to figure out how to apply a style to the line based on that though.

Is there a place I can get help finding out what API hooks I need to access etc?
Should I close this issue or leave it for now until I or someone else figures out how to do it?

**Comment by simurai** [comment]
If it's possible to add `odd` and `even` classes and just repeat the same class for wrapped lines, then the CSS would be pretty simple.

```html
<div class="line odd">Line 1</div>
<div class="line even">Line 2</div>
<div class="line even">Line 3 wrapped</div>
<div class="line even">Line 4 wrapped</div>
<div class="line odd">Line 5</div>
<div class="line odd">Line 6 wrapped</div>
<div class="line even">Line 7</div>
<div class="line odd">Line 8</div>
<div class="line even">Line 9</div>
<div class="line odd">Line 10</div>
```

```css
.line.even {
  background: silver;
}
```

See this [CodePen](http://codepen.io/simurai/pen/47bbbf127a357b46cd8fdf570f17db4c)

They could also change order in the DOM because they would be positioned absolutely. But figuring out how to add the classes might be trickier.

**Comment by kiwitotoro** [comment]
tl;dr:
How do I access the TextBuffer values from within an editor observer?
How do I find out the end point of the text file (?.getEndPosition())?
Once I have the range of the whole document, how can I loop through each line and test if it is an odd or even line from the original file?

@simurai I am currently trying to figure out how to add those classes, not knowing coffee script is a bit of an issue.

Pages like https://atom.io/docs/api/v0.165.0/TextEditor have a Buffer vs. Screen Coordinates section which explain the context.

At the moment I can access API features like this (from the decoration section https://atom.io/docs/api/v0.165.0/Decoration):

    atom.workspace.observeTextEditors (editor) ->
      range = editor.getSelectedBufferRange() 
      marker = editor.markBufferRange(range)
      decoration = editor.decorateMarker(marker, {type: 'line', class: 'zebra-stripes'})

The above code, when run, adds the class 'zebra-stripes' to all rows highlighted in the current editor window for all tabs. it uses `getSelectedBufferRange` and applies the decoration to every line in that range. So I assume all I need to do is get the range from the beginning to the end of the document...

Range (https://atom.io/docs/api/v0.165.0/Range) takes a start and end point and the TextBuffer object (https://atom.io/docs/api/v0.165.0/TextBuffer#instance-getRange) has a get range which has a ::getEndPosition attribute But I don't know how to access the TextBuffer functions (editor.getRange() causes an undefined error).

Once I have that I need to figure out how to loop through the buffer and give it a different class depending on the line number of the original text document.

I also think it would be more flexible to add a data-original-line-number="<line-number>" to the lines so I can test for odd or even values. This would allow more flexibility but decorateMarker only seems to offer class, not data adaption.

Also, I solved(hacked) the highlight not showing issue by adding a transparent background:

      .editor .line.zebra-stripes{
        background-color:rgba(255, 255, 155, 0.30);
      }

This works (as I'm using the dark theme, but I will probably need to find a way to test the background colour in the theme and add or subtract from that. (I would prefer to have the highlight above the lines but anyway...)

**Comment by postcasio** [comment]
You *can* do this with just LESS. The trick is to just look for lines where the last character of the `data-screen-row` attribute is 1, 3, 5, 7 or 9.

```less
.generate-stripes(@i: 1) when (@i <= 10) {
	atom-text-editor::shadow .lines .line[data-screen-row$="@{i}"] {
		background: rgba(0, 0, 0, 0.2);
	}

	.generate-stripes((@i + 2));
}

.generate-stripes();
```

**Comment by kiwitotoro** [comment]
That is an interesting solution and it does show the info in the correct way (at least the lines don't move around randomly :D ). If you want to add that to the stack overflow question I will accept it, but maybe add how to do it without the shadow dom enabled as well (I almost replied that it didn't work):

    .generate-stripes(@i: 0) when (@i < 10) {
      .editor .line[data-screen-row$="@{i}"] {
        background: rgba(255, 255, 255,0.02);
      }
      .generate-stripes((@i + 2));
    }
    .generate-stripes();

I will keep looking into a plugin which handles the soft wraps but as this is a solution I will close the issue.
